src/symbol/cross_tile_symbol_index.js

```diff
@@ -230,11 +255,23 @@ class CrossTileSymbolIndex {
 
         for (const tile of tiles) {
             const symbolBucket = ((tile.getBucket(styleLayer): any): SymbolBucket);
-            if (!symbolBucket) continue;
+            const bucketFeatureIndex = tile.latestFeatureIndex;
+            if (!symbolBucket || !bucketFeatureIndex) continue;
 
             if (!symbolBucket.bucketInstanceId) {
                 symbolBucket.bucketInstanceId = ++this.maxBucketInstanceId;
             }
+            // As long as this bucket is used in any committed placement, we have to hold onto
+            // the matching FeatureIndex/CollisionBoxArray/data for querying purposes
+            this.lastUsedBuckets[symbolBucket.bucketInstanceId] = true;
+            this.retainedBuckets[symbolBucket.bucketInstanceId] = new RetainedBucket(
+                symbolBucket.bucketInstanceId,
+                bucketFeatureIndex,
+                symbolBucket.sourceLayerIndex,
+                symbolBucket.index,
+                tile.tileID
+            );
+
```

Comment:
This logic is more about holding onto data during incremental placement than it is about holding onto data between frames. `lastUsedBuckets` is only cleared during the `prune` operation that happens on committing a placement. That allows us to hold onto `RetainedBuckets` that weren't used in the previous/current frame, but will still be part of the next placement that gets committed. Does that make sense?

`RetainedBuckets` is probably a misleading name, because it makes it sound like we're holding on to the buckets themselves (and thus makes it sound like something related to rendering). What we're actually holding onto is all the bucket-specific data necessary for symbol querying (basically, the raw vector tile data plus some metadata). Maybe `RetainedBucketQueryData`?

A further refactoring that might make sense is to stop holding on to the `FeatureIndex` entirely here. We're using it for shared code that loads features from the underlying vector data, but we don't need to hold onto all the data it has in its non-symbol `GridIndex`, since we always query the latest version for non-symbol features.

Reply:
🤔 Not sure if it would have a noticeable performance impact, but the current implementation of `lastUsedBuckets` holds on to more than it needs to, since it will hold onto any bucket that gets loaded during any frame of an incremental placement, even if that bucket didn't actually end up used in the placement (because of where it was in the layer order). I'll look at changing that behavior so the intent is clearer.

Reply:
Thanks for explaining, I think I had confused `prune` with `removeStaleBuckets`. This makes sense.

Renaming to `RetainedBucketQueryData` sounds good to me.

What is the advantage of loading the exact feature used for placement instead of loading features with matching `crossTileIDs`?

Reply:
Hrm, that's a good question, I haven't thought through what it would be like to implement that. So instead of storing bucket/feature indices in the `CollisionIndex`, we'd just store a `crossTileID` and maintain a reverse-lookup index to go from a `crossTileID` to the latest matching bucket/`symbolInstance`?

Initial reactions:
 - There can be multiple symbol instances per feature, and our querying logic is feature-based, so we'd have to be careful to handle duplicates/merging correctly there.
 - We'd have to figure out something reasonable to do with `crossTileID`s that were in the latest placement, but point to a tile removed from the pyramid since the placement happened.
 - Using the latest data instead of the data that was included in the latest placement opens up some room for some subtle mismatches that could in some cases be hard to debug (e.g. your query shows a symbol as being within your query geometry, but the returned `GeoJSONFeature` appears to be outside your query geometry).

I think the honest answer is "I didn't think of doing it that way". What would be the major advantage of switching to that approach? Do you think it might be a simplicity win, or are you more thinking about being able to discard data earlier?

In most cases, I don't think the current implementation leads to much extra data being retained, but the worst case is probably when there's something like a `setData`-style GeoJSON animation causing tiles with symbols to update really frequently (say, every 20ms). In that case, we'd be holding on to ~15 copies of the tile data over the course of a 300ms placement.... which sounds less-than-optimal, although I still don't think it'd be that bad (not much data, the data would have to wait for GC either way, etc.).

Reply:
I was thinking about it from a simplicity point of view (not having to deal with retaining logic), but I think those are some good points. Especially the third one. I think your current approach is 👍

Reply:
I got rid of the pruning entirely by moving `RetainedQueryData` into `Placement` -- it's a much better fit since the lifetime of the data needs to be the same as the lifetime of the `Placement`. This also avoids retaining data for buckets that never get included in a placement, which means the `setData`-animation worst case should go from ~15x tile data storage to ~2x storage.